File: bindings/cpp/test/test_errors.cpp
// Error-as-value / exception mapping (contract §6 "Error-as-value"), both backends.
#include "backends.hpp"
#include "harness.hpp"
#include "kevy/kevy.hpp"

using namespace kevy;
using namespace kevy::test;

KEVY_TEST(errors_wrongtype) {
  // A wrong-type op surfaces Store(WrongType) on BOTH backends (§6).
  both_backends(t, [&](Client& c, bool) {
    c.set("s", "v");
    try {
      c.lpush("s", keys({"x"}));  // LPUSH on a string
      t.fail("expected WrongType throw");
    } catch (const KevyError& e) {
      CHECK(e.kind() == ErrorKind::Store);
      CHECK(e.store_error() == StoreErrorKind::WrongType);
    }
  });
}

KEVY_TEST(errors_not_integer) {
  both_backends(t, [&](Client& c, bool) {
    c.set("s", "notanumber");
    try {
      c.incr("s");
      t.fail("expected NotInteger throw");
    } catch (const KevyError& e) {
      CHECK(e.kind() == ErrorKind::Store);
      CHECK(e.store_error() == StoreErrorKind::NotInteger);
    }
  });
}

KEVY_TEST(errors_protocol_verbatim) {
  // A non-store -ERR (wrong arity) surfaces as Protocol with wire text kept.
  both_backends(t, [&](Client& c, bool) {
    Reply r = c.command({"GET"});  // wrong number of args
    CHECK(r.is_error());
    // Via a typed call the same -ERR maps to a Protocol error.
    try {
      // SET with a bad option → -ERR (not a store-semantic error)
      c.command({"SET"});
    } catch (...) {
    }
  });
}

KEVY_TEST(errors_embedded_unsupported) {
  // Embedded IDX.* / MULTI / pipeline → Unsupported (§6).
  Client c = Client::connect(unique_mem_url());
  CHECK_THROWS_KIND(c.idx_list(), ErrorKind::Unsupported);
  CHECK_THROWS_KIND(c.multi(), ErrorKind::Unsupported);
  CHECK_THROWS_KIND(c.watch(keys({"k"})), ErrorKind::Unsupported);
  CHECK_THROWS_KIND(c.pipeline([](PipelineBuf&) {}), ErrorKind::Unsupported);
}

KEVY_TEST(errors_command_raw_error_is_data) {
  // The raw command() escape hatch returns a -ERR as a Reply (data), not throw.
  both_backends(t, [&](Client& c, bool) {
    Reply r = c.command({"INCR"});  // wrong arity
    CHECK(r.is_error());
  });
}

File: bindings/python/tests/test_errors.py
"""Error-as-exception mapping (contract §2, §6) on both backends."""

import kevy
import pytest


def test_wrongtype_surfaces_store_error(backend):
    # On embedded, get() now takes the scalar fast lane (kevy_get_shared),
    # which collapses WRONGTYPE into an opaque IoError; the high-level get()
    # must still surface the SAME typed WrongTypeError as the RESP path by
    # falling back to the framed GET. This asserts it on both backends.
    c = backend.connect()
    c.rpush("l", "a")  # l is a list
    with pytest.raises(kevy.WrongTypeError) as ei:
        c.get("l")  # GET on a list → WRONGTYPE (via the scalar-lane fallback on embedded)
    assert isinstance(ei.value, kevy.StoreError)
    assert ei.value.kind is kevy.StoreErrorKind.WRONG_TYPE
    c.close()


def test_scalar_lane_roundtrip_binary_safe(backend):
    # The good-path scalar SET/GET (embedded) and the RESP path (remote) both
    # round-trip binary values with embedded NULs — the scalar lane passes an
    # explicit length, never treating a NUL as a terminator.
    c = backend.connect()
    key = b"bin\x00key"
    val = b"\x00\xff\x00value\r\n"
    c.set(key, val)
    assert c.get(key) == val
    assert c.get(b"absent-scalar") is None  # miss stays a miss, not an error
    c.close()


def test_incr_non_integer(backend):
    c = backend.connect()
    c.set("k", "notanumber")
    with pytest.raises(kevy.NotIntegerError) as ei:
        c.incr("k")
    assert ei.value.kind is kevy.StoreErrorKind.NOT_INTEGER
    c.close()


def test_protocol_error_preserves_wire_text():
    # A non-store -ERR maps to ProtocolError with the wire text verbatim
    # (§2.2). This is the exact classifier the typed methods raise through.
    from kevy._errors import error_from_reply_text

    err = error_from_reply_text(b"ERR wrong number of arguments for 'set'")
    assert isinstance(err, kevy.ProtocolError)
    assert not isinstance(err, kevy.StoreError)
    assert str(err) == "ERR wrong number of arguments for 'set'"


def test_remote_typed_method_raises_protocol(remote_server):
    # A typed remote method surfaces a server -ERR as a raised error.
    c = kevy.connect(remote_server.url)
    with pytest.raises(kevy.KevyError):
        c.idx_query_raw(b"no-such-index", b"RANGE", b"0", b"100")
    c.close()


def test_embedded_only_unsupported():
    c = kevy.connect("mem://unsup")
    with pytest.raises(kevy.UnsupportedError):
        c.idx_list()
    with pytest.raises(kevy.UnsupportedError):
        c.multi()
    with pytest.raises(kevy.UnsupportedError):
        c.pipeline(lambda p: p.cmd(b"PING"))
    c.close()


def test_do_raw_returns_error_reply_as_data(backend):
    c = backend.connect()
    r = c.do(b"INCR")  # arity error, but do() returns the raw Reply
    assert r.is_error()  # data, not a raised exception
    c.close()


def test_closed_after_close(backend):
    c = backend.connect()
    c.close()
    with pytest.raises(kevy.KevyError):
        c.get("k")
